lib/SILOptimizer/Utils/OptimizerStatsUtils.cpp

```diff
@@ -0,0 +1,970 @@
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+///
+/// This file implements collection and dumping of statistics about SILModules,
+/// SILFunctions and memory consumption during the execution of SIL
+/// optimization pipelines.
+///
+/// The following statistics are collected:
+/// - For SILFunctions: the number of SIL basic blocks, the number of SIL
+/// instructions
+///
+/// - For SILModules: the number of SIL basic blocks, the number of SIL
+/// instructions, the number of SILFunctions, the amount of memory used by the
+/// compiler.
+///
+/// By default, any collection of statistics is disabled to avoid affecting
+/// compile times.
+///
+/// One can enable the collection of statistics and dumping of these statistics
+/// for the whole SILModule and/or for SILFunctions.
+///
+/// To reduce the amount of produced data, one can set thresholds in such a way
+/// that changes in the statistics are only reported if the delta between the
+/// old and the new values are at least X%. The deltas are computed using the
+/// following formula:
+///  Delta = (NewValue - OldValue) / OldValue
+///
+/// Thresholds provide a simple way to perform a simple filtering of the
+/// collected statistics during the compilation. But if there is a need for a
+/// more complex analysis of collected data (e.g. aggregation by a pipeline
+/// stage or by the type of a transformation), it is often better to dump
+/// as much data as possible into a file using e.g. `-sil-stats-dump-all
+/// -sil-stats-modules -sil-stats-functions` and then e.g. use the helper scripts
+/// to store the collected data into a database and then perform complex queries
+/// on it. Many kinds of analysis can be then formulated pretty easily as SQL
+/// queries.
+///
+/// The output format is a set of CSV (comma separated values) lines. Each lines
+/// represents one counter or one counter change.
+///
+/// For counter updates it looks like this:
+///  Kind, CounterName, StageName, TransformName, TransformPassNumber,
+///  DeltaValue, OldCounterValue, NewCounterValue, Duration, Symbol
+///
+/// For counter updates it looks like this:
+///  Kind, CounterName, StageName, TransformName, TransformPassNumber,
+///  CounterValue, Duration, Symbol
+///
+/// where Kind is one of "function", "module", "function_history",
+///       CounterName is one of "block", "inst", "function", "memory",
+///       Symbol is e.g. the name of a function.
+///       StageName and TransformName are the names of the current optimizer
+///       pipeline stage and current transform.
+///       Duration is the duration of the transformation.
+//===----------------------------------------------------------------------===//
+
+#include <iostream>
```

Comment:
This violates LLVM coding style: http://llvm.org/docs/CodingStandards.html#include-iostream-is-forbidden

You should be using llvm::errs() below.